GitUI/CommitInfo/RefsFormatter.cs

```diff
@@ -0,0 +1,116 @@
+            bool allowLocal = AppSettings.CommitInfoShowContainedInBranchesLocal;
+            bool allowRemote = getRemote;
+
+            foreach (var branch in branches)
+            {
+                string noPrefixBranch = branch;
+                bool branchIsLocal;
+                if (getLocal && getRemote)
+                {
+                    // "git branch -a" prefixes remote branches with "remotes/"
+                    // It is possible to create a local branch named "remotes/origin/something"
+                    // so this check is not 100% reliable.
+                    // This shouldn't be a big problem if we're only displaying information.
+                    // This could be solved by listing local and remote branches separately.
+                    branchIsLocal = !branch.StartsWith(remotesPrefix);
+                    if (!branchIsLocal)
+                    {
+                        noPrefixBranch = branch.Substring(remotesPrefix.Length);
+                    }
+                }
+                else
+                {
+                    branchIsLocal = !getRemote;
+                }
+
+                if ((branchIsLocal && allowLocal) || (!branchIsLocal && allowRemote))
+                {
+                    var branchText = showAsLinks
+                        ? LinkFactory.CreateBranchLink(noPrefixBranch)
+                        : WebUtility.HtmlEncode(noPrefixBranch);
+
+                    if (limit && links.Count == MaximumDisplayedRefsIfLimited + 2)
+                    {
+                        links.RemoveRange(MaximumDisplayedRefsIfLimited, 2);
+                        truncated = true;
+                        break; // from foreach
+                    }
+
+                    links.Add(branchText);
+                }
+
+                if (branchIsLocal && AppSettings.CommitInfoShowContainedInBranchesRemoteIfNoLocal)
+                {
+                    allowRemote = false;
+                }
+            }
+
+            return ToString(links, Strings.ContainedInBranches, Strings.ContainedInNoBranch, "branches", truncated);
+        }
+
+        public static string FormatTags(IReadOnlyList<string> tags, bool showAsLinks, bool limit)
+        {
+            bool truncate = limit && tags.Count > MaximumDisplayedRefsIfLimited + 2;
+            var links = FormatTags(truncate ? tags.Take(MaximumDisplayedRefsIfLimited) : tags);
+            return ToString(links, Strings.ContainedInTags, Strings.ContainedInNoTag, "tags", truncate);
+
+            IEnumerable<string> FormatTags(IEnumerable<string> tags_)
+            {
+                return tags_.Select(s => showAsLinks ? LinkFactory.CreateTagLink(s) : WebUtility.HtmlEncode(s));
+            }
+        }
+
+        private static string ToString(IEnumerable<string> links, string prefix, string textIfEmpty, string refsType, bool truncated)
+        {
+            if (links.Any())
```

Comment:
❗️ Multiple enumerations

```cs
        private static string ToString(IEnumerable<string> links, string prefix, string textIfEmpty, string refsType, bool truncated)
        {
            var links = links.Join(Environment.NewLine);
            if (links.Length < 1)
                return string.Empty;
            // else format as necessary
```

Reply:
(Resharper works for me - thank you for the license!) I saw this warning, too. Though in this case, Re# is not perfect IMO.  I think it is cheaper to start the enumeration twice than to allocate an array (or a string in your quite efficient proposal) - because the first enumeration is stopped at once for `Any`.
Your proposal is mixing the positive with the negative case at the expense of readability.
So: How expensive is the second enumeration for `Any` really?

Reply:
No, it is not so much the issue with the `Any` call, but the fact that the underlying collection may be enumerated multiple times when `links.Any()` is called followed by `links.Join(...)`

If the `links` underlying collection is materialised (e.g. a concrete `List<string>` or `Array`) then there will be no further allocations.
However imaging you have a complex query that fetches data from a database, a file or a network stream, and the result hasn't materialised:
```cs
var links = LoadInfoFromDatabase(ids)
				.Union(FetchLinksFromHttp)
				.Where(id => id < 100);

var formattedLinks = ToString(links, ...);
```

When the code reaches `Any` it hasn't loaded anything from a db or hit a remote endpoint. Executing `Any` will perform that so we have a result - "have results or not?".
Then `Join` operation will **again** execute the db request and hit the remote endpoint.